src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,127 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizer;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.CostModel;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	protected final StoppingConditionForIterativeImprovement condition;
+
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt, final StoppingConditionForIterativeImprovement x) {
+		assert ctxt != null;
+		assert x != null;
+		context = ctxt;
+		condition = x;
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+		return optimize( context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false) );
+	}
+
+	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
+		// The best plan we have found so far. As we have only found one plan, it is the best one so far.
+		PhysicalPlan bestPlan = initialPlan;
+		PhysicalPlan currentPlan = bestPlan; // This variable will hold the plan which is currently being worked on.
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel();
+		CompletableFuture<Double> futureCost = costModel.initiateCostEstimation(bestPlan);

```

Comment:
getEstimates (at least the one I have in this branch) only takes a list and spits out an array. In this case I just want a single value for a single plan. Wouldn't it be a waste to make a function call as well as have the additional overhead of the two containers just to get. Addtionally, as far as I understand it we will lose parallelism by seeking to call the get() function immediately after making the completablefuture rather than processing other things in the meanwhile if we can.

Are there any performance upsides to using the getEstimates function here?

Reply:
> getEstimates (at least the one I have in this branch) only takes a list and spits out an array. In this case I just want a single value for a single plan. Wouldn't it be a waste to make a function call as well as have the additional overhead of the two containers just to get.

Don't worry so much about such low-level performance optimization opportunities. The Java runtime is usually doing a reasonably good job with many of these things.

By the way, you do not have to create the list for getEstimates yourself. Instead, there is a version of this function that is declared as follows:
```
public static Double[] getEstimates( final CostModel costModel,
                                     final PhysicalPlan... plans )
```
The type `PhysicalPlan...` means that you can call the function with as many PhysicalPlan objects as you like, which will then be treated as an array. So, you can call
```
getEstimates(myCostModel, plan1, plan2, plan3);
```
but you can also simply call

```
getEstimates(myCostModel, plan1);
```

> Addtionally, as far as I understand it we will lose parallelism by seeking to call the get() function immediately after making the completablefuture rather than processing other things in the meanwhile if we can.

That's a good point in general, but these few things that you had in between wouldn't make much of a difference I would assume.

An alternative idea the may lead to an actual performance gain would be to unroll the outer loop by pulling out the first iteration step of that loop and do that step first, separate from the rest of the loop. Then, in that first step, you can request the cost of the initial plan together with requesting the cost of the neighbors of that plan. However, since that would make the code a bit more complicated, I would suggest you ignore this idea for the moment.

> Are there any performance upsides to using the getEstimates function here?

No. However, it will make the code significantly easier to read.

Reply:
Using
Double initialCost = CostEstimationUtils.getEstimates(costModel, initialPlan);
instead of my current implementation gives me an error in the IDE telling me that I need to make initialCost into an array or make Double the return type of getEstimates.

I'm making initialCost as an array with a single member, this should hopefully only lose negligible performance since an array only has a few bytes worth of overhead in C and probably something similar in Java.

Reply:
Yes, the return type of `getEstimates` is still an array, even if you pass only one `PhysicalPlan` to this function.